components/spinner.py

```diff
@@ -3,29 +3,131 @@
 
-    def on_enable(self):
+    WHEEL_SPIN_FACTOR = 0.3
+
+    def __init__(self) -> None:
+        self.rotations = 0
+        self.state = None
+        self.lastCol = None
+        self.required_colour = None
+
+    def on_enable(self) -> None:
         self.spinner_motor.stopMotor()
-        self.spinner_solenoid.set(wpilib.DoubleSolenoid.Value.kReverse)
+        self.spinner_solenoid.set(True)
+        self.rotation = 0
+        self.state = None
+        self.lastCol = None
+        self.required_colour = None
+
+    # Methods to be called by controller
+    def is_complete(self) -> bool:
+        return self.state == None
+
+    def do_rotation_control(self) -> None:  # to start the rotation routine
+        self.state = "rotation"
+        self.rotations = 0
+        self.lastCol = None
+        self.piston_down()
 
-    def execute(self):
-        pass
+    def go_to_colour(self, colour: str) -> None:  # to start the position routine
+        self.state = "position"
+        self.required_colour = colour
+        self.piston_down()
 
-    def read_colour(self):
+    def piston_up(self) -> None:  # as functions to make swapping easier
+        self.spinner_solenoid.set(True)
+
+    def piston_down(self) -> None:
+        self.spinner_solenoid.set(False)
+
+    def read_colour(self) -> tuple:
         return (self.colour_sensor.getColor(), self.colour_sensor.getIR())
 
-    def piston_up(self):
-        self.spinner_solenoid.set(wpilib.DoubleSolenoid.Value.kForward)
+    def get_wheel_dist(
+        self,
+    ) -> int:  # finds the colour wheels distance from required colour in segments
+        current_colour = self.get_current_colour()
+        if current_colour == None:
+            return 0
+        distance = (
+            self.WHEEL_ORDER.index(self.required_colour)
+            + self.COLOUR_OFFSET
+            - self.WHEEL_ORDER.index(current_colour)
+        )
+        if distance > 2:
+            return distance - 4
+        return distance
+
+    def get_current_colour(
+        self,
+    ) -> float:  # gets colour as a letter ("R", "Y", "B", "G")
+        sensed_colour, _ = self.read_colour()
+
+        distances = {}
+        for col_name, col_value in self.WHEEL_COLOURS.items():
+            distances[col_name] = col_value.dist(sensed_colour)
+
+        if (
+            distances[min(distances, key=lambda col: distances[col])]
+            < self.MAX_COLOUR_DIST
+        ):
+            return min(distances, key=lambda col: distances[col])
+
+    def run_rotation(self) -> None:  # rotation control to be called every 20ms
+        self.spinner_motor.set(
+            1 - self.rotations / 6
+        )  # speed slows down as it gets closer to required revolutions
+        current_colour = self.get_current_colour()
+        if self.lastCol != current_colour:
+            self.rotations += 1 / 8
+        self.lastCol = current_colour
+        if self.rotations >= 3.5 and self.read_colour()[1] > 35:
```

Comment:
Consider using the piston state or a variable to track whether the mechanism is up or down, if you keep this method add a comment to explain what you are doing and why as it is currently not obvious

Reply:
I made the suggested changes